bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/services/seo/LanguageNavigationSiteRootSelectionStrategy.java

```diff
@@ -0,0 +1,146 @@
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+
+import com.adobe.aem.wcm.seo.localization.SiteRootSelectionStrategy;
+import com.adobe.cq.wcm.core.components.models.ExperienceFragment;
+import com.adobe.cq.wcm.core.components.models.LanguageNavigation;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.TemplatedResource;
+import com.day.cq.wcm.api.designer.Designer;
+import com.day.cq.wcm.api.designer.Style;
+
+/**
+ * An implementation of {@link SiteRootSelectionStrategy} that looks for a language navigation component on the given page and uses it's
+ * configured navigation root as site root.
+ * <p>
+ * This Component must be explicitly enabled by providing an (empty) configuration for it. This should be the case for sites the use the
+ * language navigation core component. For any other case this Component should be kept disabled.
+ */
+@Component(
+    configurationPolicy = ConfigurationPolicy.REQUIRE,
+    property = {
+        Constants.SERVICE_RANKING + ":Integer=100"
+    },
+    service = { SiteRootSelectionStrategy.class }
+)
+public class LanguageNavigationSiteRootSelectionStrategy implements SiteRootSelectionStrategy {
+
+    @Reference(
+        cardinality = ReferenceCardinality.OPTIONAL,
+        policyOption = ReferencePolicyOption.GREEDY,
+        target = "(objectClass=*DefaultSiteRootSelectionStrategy)")
+    private SiteRootSelectionStrategy defaultSelectionStrategy;
+
+    @Override
+    @Nullable
+    public Page getSiteRoot(@NotNull Page page) {
+        Resource languageNavigation = findLanguageNavigation(page.getContentResource());
+        if (languageNavigation != null) {
+            String siteRoot = getPropertyLazy(languageNavigation, LanguageNavigation.PN_NAVIGATION_ROOT, String.class);
+            if (siteRoot != null) {
+                return page.getPageManager().getPage(siteRoot);
+            }
+        }
+        return defaultSelectionStrategy != null ? defaultSelectionStrategy.getSiteRoot(page) : null;
+    }
+
+    @Override
+    public int getStructuralDepth(@NotNull Page page) {
+        Resource languageNavigation = findLanguageNavigation(page.getContentResource());
+        if (languageNavigation != null) {
+            Integer structureDepth = getPropertyLazy(languageNavigation, LanguageNavigation.PN_STRUCTURE_DEPTH, Integer.class);
+            if (structureDepth != null) {
+                return structureDepth;
+            }
+        }
+        return defaultSelectionStrategy != null ? defaultSelectionStrategy.getStructuralDepth(page) : 1;
+    }
+
+    private Resource findLanguageNavigation(Resource contentResource) {
+        Resource templatedResource = contentResource.adaptTo(TemplatedResource.class);
```

Comment:
At some point it would make sense to have some sort of caching as this is potentially expensive performance-wise.

Reply:
True, for a start it would be good to cache the returned language navigation for the given `Page` so that when it is called more than once with the same object it does not traverse again. I will use https://guava.dev/releases/18.0/api/docs/com/google/common/cache/CacheBuilder.html#weakKeys() to build a short lived cache.